src/ImageSharp/Configuration.cs

```diff
@@ -1,272 +1,244 @@
-        /// </summary>
-        /// <param name="format">The format to discover</param>
-        /// <returns>The <see cref="IImageEncoder"/> if found otherwise null</returns>
-        public IImageEncoder FindEncoder(IImageFormat format)
-        {
-            Guard.NotNull(format, nameof(format));
-            if (this.mimeTypeEncoders.TryGetValue(format, out IImageEncoder encoder))
-            {
-                return encoder;
-            }
-
-            return null;
-        }
-
-        /// <summary>
-        /// Creates the default instance with the following <see cref="IConfigurationModule"/>s preregistered:
-        /// <para><see cref="PngConfigurationModule"/></para>
-        /// <para><see cref="JpegConfigurationModule"/></para>
-        /// <para><see cref="GifConfigurationModule"/></para>
-        /// <para><see cref="BmpConfigurationModule"/></para>
-        /// </summary>
-        /// <returns>The default configuration of <see cref="Configuration"/></returns>
-        internal static Configuration CreateDefaultInstance()
-        {
-            return new Configuration(
-                new PngConfigurationModule(),
-                new JpegConfigurationModule(),
-                new GifConfigurationModule(),
-                new BmpConfigurationModule());
-        }
-
-        /// <summary>
-        /// Sets the max header size.
-        /// </summary>
-        private void SetMaxHeaderSize()
-        {
-            this.MaxHeaderSize = this.imageFormatDetectors.Max(x => x.HeaderSize);
-        }
-    }
-}
+// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using SixLabors.ImageSharp.Formats;
+using SixLabors.ImageSharp.Formats.Bmp;
+using SixLabors.ImageSharp.Formats.Gif;
+using SixLabors.ImageSharp.Formats.Jpeg;
+using SixLabors.ImageSharp.Formats.Png;
+using SixLabors.ImageSharp.IO;
+using SixLabors.ImageSharp.Memory;
+
+namespace SixLabors.ImageSharp
+{
+    /// <summary>
+    /// Provides initialization code which allows extending the library.
+    /// </summary>
+    public sealed class Configuration
+    {
+        /// <summary>
+        /// A lazily initialized configuration default instance.
+        /// </summary>
+        private static readonly Lazy<Configuration> Lazy = new Lazy<Configuration>(CreateDefaultInstance);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Configuration" /> class.
+        /// </summary>
+        public Configuration()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Configuration"/> class.
+        /// </summary>
+        /// <param name="configuration">A configuration instance to be copied</param>
+        public Configuration(Configuration configuration)

```

Comment:
Normally I would expect this kind of constructor to do a deep copy. Wouldn't it be better to have a `Configuration.ShallowCopy()` method instead to be more explicit?

Reply:
I think the most common use is going to be create a copy of the Default configuration, that keeps all the internal objects shared with Default, and then change exactly what you want to change.

By doing a deep copy you loose the feature of sharing underlaying components between configurations. Sharing components can be really convenient for MemoryManager and ImageFormatManager; for example, we can have a derived configuration with a custom memory manager, but image formats shared with default, so if a new format is added later at runtime, the derived configurations automatically pick that new format.

Reply:
>ImageFormatManager; for example, we can have a derived configuration with a custom memory manager, but image formats shared with default, so if a new format is added later at runtime, the derived configurations automatically pick that new format.

But what happens if someone want a more restrictive list in a separate location? It would then be a lot of work creating and populating the new manager.

Reply:
@JimBobSquarePants then what it needs a deep copy is MemoryManager and ImageFormatManager.

In the case of ImageFormatManager, should it also deep copy the formats themselves?

Reply:
I would expect so yeah. .NET needs a `DeepClone()` method to make this all easier!

Reply:
@vpenades I suggested a `ShallowCopy()` method, not a deep copy. I just don't like a constructor doing a shallow copy implicitly. Being more verbose here is better IMO:
```C#
Configuration myConfig = Configuration.Default.ShallowCopy();
myConfig.MemoryManager = ...;
```